Enhancement: option to Get All Tokens with a single Blockchain Document Loader call

saginawj opened the pull request:
The Blockchain Document Loader's default behavior is to return 100 tokens at a time which is the Alchemy API limit.  The Document Loader exposes a startToken that can be used for pagination against the API.

This enhancement includes an optional get_all_tokens param (default: False) which will:

- Iterate over the Alchemy API until it receives all the tokens, and return the tokens in a single call to the loader.
- Manage all/most tokenId formats (this can be int, hex16 with zero or all the leading zeros).  There aren't constraints as to how smart contracts can represent this value, but these three are most common.

Note that a contract with 10,000 tokens will issue 100 calls to the Alchemy API, and could take about a minute, which is why this param will default to False.  But I've been using the doc loader with these utilities on the side, so figured it might make sense to build them in for others to use.

saginawj committed:
get_all_nfts implemented and tested

saginawj committed:
fixed linter error

saginawj committed:
2nd linter try given local issue

saginawj commented on langchain/document_loaders/blockchain.py:
```diff
@@ -55,34 +61,77 @@ def __init__(
+                }
+                result.append(Document(page_content=content, metadata=metadata))
+
+            if not self.get_all_tokens:
```
@skcoirz I appreciate the feedback!  I'm certainly open to suggestions on how to improve the flow of logic here.  The reason I put this self.get_all_tokens check after the first API call (which I agree is maybe a bit counter-intuitive) is because the tokenId can take many formats (e.g. int, hex with leading zeros, etc.) and that format is needed for the 'get_all' call (else it will return an empty array).  So I execute the first API call, and if self.get_all_tokens is true, I determine the tokenId type and use it to paginate through the API.  With either flow, that first call is needed.  If the flag is false, the loop stops after the first call but if true it will continue until no tokens are returned from the API.  

I hope that's clear.  I can certainly change this flow if the logic is unclear.  I just wanted to outline the thinking behind my thought process. Thanks again!

skcoirz replied:
I see your point. The first call is shared by both. Even we split the logic after the first call, we still need to make the url call later for get_all_tokens case, which make the code redundant a little bit as the same function is used in 2 sequential places. So like what you did here, a `do, while` makes sense.

I’m thinking of your needs here + our chat of time consumption. Have you considered async calls? We can make the url calls on parallel. I shared my thoughts with Codeium. The following is the sponsored example. :) Please take a look if this makes it more efficient. Thanks!

```python3
import asyncio
import aiohttp
from typing import List
from .document import Document

async def afetch(session: aiohttp.ClientSession, url: str) -> dict:
    async with session.get(url) as response:
        if response.status != 200:
            raise ValueError(f"Request failed with status code {response.status}")
        return await response.json()

async def afetch_all(session: aiohttp.ClientSession, urls: List[str]) -> List[List[dict]]:
    results = await asyncio.gather(*[afetch(session, url) for url in urls])
    return [result["nfts"] for result in results if result["nfts"]]

# TODO: following the convention in this code base, it’s probably better to create another 
# function with name `aload` and handle all the downstream calls in all `a-` prefix 
# functions in the dependent classes.
async def aload(self, max_requests: int = 100) -> List[Document]:
    result = []
    current_start_token = self.startToken
    urls = []
    request_count = 0

    # TODO: I found you don’t have an exit condition check here. AI also failed to detect it. 
    # (very interesting.. clearly it’s a chat bot instead of an analytic bot…) 
    # I’m not really familiar with the use case here, do you have a good suggestion what 
    # can be used as a stricter exit condition? Thanks!
    while request_count < max_requests:
        url = (
            f"https://{self.blockchainType}.g.alchemy.com/nft/v2/"
            f"{self.api_key}/getNFTsForCollection?withMetadata="
            f"True&contractAddress={self.contract_address}"
            f"&startToken={current_start_token}"
        )
        urls.append(url)

        if not self.get_all_tokens:
            break

        current_start_token = self._get_next_tokenId(result[-1].metadata["tokenId"]) if result else self.startToken

        request_count += 1

    async with aiohttp.ClientSession() as session:
        nft_lists = await afetch_all(session, urls)

        for nfts in nft_lists:
            for item in nfts:
                content = str(item)
                tokenId = item["id"]["tokenId"]
                metadata = {
                    "source": self.contract_address,
                    "blockchain": self.blockchainType,
                    "tokenId": tokenId,
                }
                result.append(Document(page_content=content, metadata=metadata))

    if not result:
        raise ValueError(f"No NFTs found for contract address {self.contract_address}")

    return result
```

skcoirz replied:
I see your point. The first call is shared by both. Even we split the logic after the first call, we still need to make the url call later for get_all_tokens case, which make the code redundant a little bit as the same function is used in 2 sequential places. So like what you did here, a `do while` makes sense.

saginawj replied:
Sounds good @skcoirz!  Shall I mark this convo as resolved?

skcoirz replied:
sounds good!

saginawj commented on langchain/document_loaders/blockchain.py:
```diff
@@ -55,34 +61,77 @@ def __init__(
+        else:
+            return str(result)
+
+    def _detect_value_type(self, tokenId: str) -> str:
```
thanks @skcoirz  :).  will fix in next commit

saginawj commented on langchain/document_loaders/blockchain.py:
```diff
@@ -55,34 +61,77 @@ def __init__(
+
+        current_start_token = self.startToken
+
+        while True:
```
Open to the concept- but want to make sure I understand your thinking.  Are you suggesting we exit the loop if the set time is exceeded, or only add the time check for testing purposes?

skcoirz replied:
Yeah I think it would be convenient to provide users a way to early stop based on time check. This could help us mitigate very extreme cases when the expected time cost is very high.

A similar idea is in use in `AgentExecutor` where a single user request may requires multiple rounds of immediate searching. Code pointer below. I think it’s a good idea. Maybe we can consider something similar here?

https://github.com/hwchase17/langchain/blob/master/langchain/agents/agent.py#L830

saginawj replied:
@skcoirz I took a look at the asyncio process and impelemented a general asyncio-based approach in a new blockchain_async file that I intend to remove from my next commit but keeping here for review.  This approach must have a glitch as it still takes 45 seconds to load, or it reaches api rate limits on the api.

I updated blockchain.py doc loader with a simple check against a new max_execution_time param that will exit the flow if this number is exceeded.  I believe this is a simple approach that can accomplish the goal, but happy to continue trying the asyncio model as well.

saginawj committed:
set util method to static

saginawj committed:
max_execution_time param added to support long-running calls

@skcoirz for a simple solution for get_all_tokens, I added a timeout param.  I also added an asyncio file that attempts to address this asynchronously.  Note that I tried a version of this that kept failing due to rate limit issues against the API, and the current implementation must have something wrong as it returns the right amount of NFTs but in the same amount of time as the sync call.  I include for your review but intend to either throw this version out, or improve it and replace the sync.  But wanted to first show the simple fix.

skcoirz commented on langchain/document_loaders/blockchain_async.py:
```diff
@@ -0,0 +1,172 @@
+                        }
+                        result.append(Document(page_content=content, metadata=metadata))
+
+                    next_start_token = self._get_next_tokenId(
+                        result[-1].metadata["tokenId"]
+                    )
```
I think this is the blocker of this async idea. Each call has the dependency with the one before... This is also why I removed my previous comment with the AI-generated example... (in case you have received the email notification even though I had removed it...)
From my limited understanding, I think for blockchain async, it's possible when we are fetching with multiple heads of blockchains so that we can process them on parallel, but for each chain, it's a sequential process. But my knowledge of blockchain is limited. The thought above is based on our example logic here. (I also did some exploration online for `Ethereum` and `alchemy api`, but couldn't find a good solution :<)

saginawj replied:
thanks for the feedback.  I'll take another look at this, but unfortunately not until later this week.   The good news is that I think there are 3 approaches:

- the one currently in blockchain.py uses a max_execution_time param that the client can set if 40secs is too long to wait.  I think from a client perspective, this will be ok for most people.
- fixing the async.  currently, it is 'working' in that the test cases are passing, but it is not executing in parallel correctly.  so the request takes too long.
- at some point today, I had the solution 'working' but the requests were failing because they were being sent to the API too quickly and I was hitting rate limits.  Unfortunately, I moved off that approach too quickly and need to remember how I had this setup.  There must be a way to batch or slow-down those requests to prevent the rate limit issue.

saginawj replied:
@skcoirz ok so I think the async is pretty close.  I ran into a few issues with rate limits in the API, then others with representing the correct url value dynamically (we don't know how many async tasks there will be ahead of time). 

I won't have many cycles for a few days to finalize the async work, but i kept a version of it on my fork [here ](https://github.com/saginawj/langchain/blob/get-all-nfts-async/langchain/document_loaders/blockchain_async.py).  In the PR, i updated the sync version so it is working with the timeout, and I believe this model will work for most use cases, though I do want to figure out how to optimize with the async calls.  I also don't want to overcomplicated the doc loader or implement in a way that is not aligned with the other loader, so any feedback here would be appreciated!   I've been through several iterations on the async (e.g. with various strategies for dealing with rate limits on the API), but scaled back in the posted version.

skcoirz replied:
yeah sounds good. async is not trivial. We don’t need to rush for it. Thank you for your patience and support in the iterations and discussions here! I really appreciate it!

saginawj committed:
added timeout, removed asnyc for now, will introduce in separate branch

saginawj committed:
linter fix in test case

saginawj committed:
black reformat

hwchase17 commented:
thanks! and thanks @skcoirz for review!
